wordpress-mobile/WordPress-iOS
# Add device modes for View Site previews

### Description

This PR adds device modes to the View Site preview screen.

I utilized the existing `PreviewWebKitViewController`, which subclasses `WebKitViewController`. I also added an optional boolean (default false) to the factory method of the parent class. Since the site preview isn't associated with a particular `post`, it was necessary to make that optional in the subclass.

I''ve added a button for the device modes here: https://github.com/wordpress-mobile/WordPress-iOS/pull/15758/files#diff-8515b18aa430fd6d7e25e540843b387c5b0ef97acdaedf383c39e7b653c235d9R125-R127, but I'm not sure whether I've tested all cases for when `linkBehavior == .all`. @bjtitus :wave: :smile: , do you know whether there are cases where this subclass is being used where the setting is left unimplemented? . Fwiw, when I tried making `post` private, I noticed a build failure from a required visibility in `SearchManager`, so I tried a viewing a few posts from the post list search screen, and everything I tried seems to work as expected.

Ideally, we could enable the device modes everywhere we use these webviews for previews, for consistency, so please feel free to let me know if I've missed a case. On a related note, I'd considered moving the device modes to the parent class, but I think it makes sense to defer to a later PR, as I'd want to have a more comprehensive understanding of the impact before introducing it there (there may be some cases / paths where the feature does not make sense, for example). Wdyt?

To test:

* On the My Site screen, scoll to the bottom and under "External" tap "View Site"
* When the preview opens, expect a device modes button to appear in the lower right corner
* Tap the device modes button (the currently selected device should reflect the default for the device you are on (i.e. mobile or tablet)
* Tap a different device mode (e.g. desktop)
* Expect the preview to reload with the selected device mode's viewport size

PR submission checklist:
